### ISSUE TITLE
Download Chrome automatically

### ISSUE THREAD

**Issue by minusnine** [tebeka/selenium#55]
There is a TODO in init.go to download Chrome automatically.

For my notes, there are continuous builds in this GCS bucket:

http://storage.googleapis.com/chromium-browser-continuous?prefix=Linux_x64

**Comment by RaviTezu** [comment]
Hi @minusnine 
I would like do this ToDo task, can you please give me more details on how it should be done or what is the expectation? 

>> Download chrome automatically. 
Do you mean there should be a init.go file inside the chrome directory and it should download the chrome driver automatically? Thanks in advance.

**Comment by minusnine** [comment]
Hi Ravi, thanks for volunteering.

[init.go](https://github.com/tebeka/selenium/blob/master/vendor/init.go) is a simple command line utility that downloads a list of files, checks their checksums, and unpacks them. It should be changed to also download Chrome automatically.

To do this, it should use the Google Cloud Storage API, likely through the [official Go API](https://godoc.org/cloud.google.com/go/storage#BucketHandle.Objects), to iterate through the [continuous build bucket for Chromium](http://storage.googleapis.com/chromium-browser-continuous?prefix=Linux_x64) to find the newest available binary, and then download and unpack it.

Please let me know if you have any other questions.

**Comment by RaviTezu** [comment]
Hi @minusnine 
I wrote this simple script just to tryout the download: https://gist.github.com/RaviTezu/b8607ff3e0c3da176df89cb8b1c315ec

The problem I see here is, I have to create service account(using my personal gmail account) and use its credentials to create a storage client like [this](https://gist.github.com/RaviTezu/b8607ff3e0c3da176df89cb8b1c315ec#file-chromedownloader-go-L32). I have also checked the sample avaible [here](https://github.com/GoogleCloudPlatform/golang-samples/tree/master/storage) but creating a client needs the projectID, but I am not able to know the Project ID of `chromium-browser-continuous` bucket. 

In short, I need a way to create a storage client with some common service account or someother way to create a client and browse the contents of `chromium-browser-continuous` bucket. Do you have any suggestions here? Thanks in advance. 

I still have to do the following: 
1. Refactor the variable names and also, integrate this code into init.go. 
2. Perform the checksum once the file is downloaded.

**Comment by minusnine** [comment]
There must be a way to access the data anonymously, since I can fetch a file constructed from the bucket manifest directly using an incognito browser window.

What happens if you don't call `option.WithServiceAccountFile` and only call `storage.NewClient(ctx)`?

**Comment by RaviTezu** [comment]
Hi @minusnine 
Thanks for the reply. 

Without `option.WithServiceAccountFile` I get a could not find default credentials error as below: 

```
ERROR: logging before flag.Parse: I0502 00:04:21.992061   18392 main.go:34] Cannot create a storage client: dialing: google.DefaultTokenSource: google: could not find default credentials. See https://developers
.google.com/accounts/docs/application-default-credentials for more information.
panic: runtime error: invalid memory address or nil pointer dereference
```

As you said, I am trying to see, if we can create a client without a creds.

**Comment by minusnine** [comment]
The docs don't officially say how to access public data through the API:
https://cloud.google.com/storage/docs/access-public-data

The docs imply that setting an API key is required:
https://cloud.google.com/vision/docs/common/auth#set_up_an_api_key

By looking through the client source, I was able to determine that we can shortcut that error message by setting a dummy API key:

client, err := storage.NewClient(ctx, option.WithAPIKey("key-not-used-with-public-data"))

That's not obvious. I'll file a bug to improve the documentation with something, though I suspect this is not exactly supported, even if it works. I think it's good enough for us, for now, though.

Let me know if that doesn't work for you.

**Comment by minusnine** [comment]
Ah, actually, the right thing to do is to use:
```go
client, err := storage.NewClient(ctx, option.WithHTTPClient(http.DefaultClient))
```

**Comment by RaviTezu** [comment]
Hi @minusnine 
Awesome. I have submitted the [PR](https://github.com/tebeka/selenium/pull/58) here.

**Comment by minusnine** [comment]
Merged. Thanks again!

Might you be interested in updating remote_test.go to point to the downloaded file by default?

**Comment by RaviTezu** [comment]
Thanks @minusnine 
I will start working on remote_test.go :)

**Comment by RaviTezu** [comment]
Hi @minusnine 

I have update the remote_test.go to point to the downloaded file and raise this [PR](https://github.com/tebeka/selenium/pull/64).
However, I just found that the tests are failing becase the latest chromium according to the bucket-list [here](https://console.cloud.google.com/storage/browser/chromium-browser-continuous/Linux_x64/?pli=1) is `Chromium 51.0.2684.0`. But the actual latest release is something like 58.0. 
The tests are failing with the below error: 

```
unknown error - 33: session not created exception: Chrome version must be >= 56.0.2884.0
```
I have tried downloading the latest chromium from here and https://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html?prefix=Linux_x64/473420/ and pointed the remote_test.go to this location and all the tests passed. 

I am suspecting storage API we are using is not being up to date.

**Comment by minusnine** [comment]
Can you update init.go to this bucket: chromium-browser-snapshots ?

https://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html?prefix=Linux_x64/

**Comment by RaviTezu** [comment]
Hi @minusnine 
I have updated the bucket from chromium-browser-continuous to chromium-browser-snapshots and I see the latest version is `Chromium 60.0.3110.0 ` and I have executed the tests and all passed. 
Added a commit before the existing commit in the same PR.

**Comment by minusnine** [comment]
Merged! Thanks for the contribution! Please let me know if you are interested in contributing more, or just send PRs.
